src/custom_hdrs.c

```diff
@@ -0,0 +1,95 @@
+/**
+* @file src/custom_hdrs.c  Custom headers control
+*
+* Copyright (C) 2010 Creytiv.com
+*/
+#include <re.h>
+#include <baresip.h>
+#include "core.h"
+
+static void hdr_destructor(void *arg)
+{
+	struct sip_hdr *hdr = arg;
+
+	mem_deref((char *)hdr->name.p);
+	mem_deref((char *)hdr->val.p);
+}
+
+static void custom_hdrs_destructor(void *arg)
+{
+	struct list *hdrs = arg;
+	list_flush(hdrs);
+}
+
+int custom_hdrs_alloc(struct list **hdrs)
+{
+	struct list *h;
+	h = mem_zalloc(sizeof(*h), custom_hdrs_destructor);
+	if (!h)
+		return ENOMEM;
+	list_init(h);
+	*hdrs = h;
+	return 0;
+}
```

Comment:
I am wondering if it is better to use a compound list instead of allocated list.

Most of the code in baresip and libre using a list is declared like this:

```
struct foo {
    struct list my_list;
};
```

this way there is no need for an allocation. The list is empty by default.

To initialize the list:

`list_init();`

To destroy the list members:

`list_flush();`

To check if the list is empty or have any members:

`list_isempty()` or `list.head != NULL`

if you write the code like this you will find that it becomes simpler.
but it is not possible to use `mem_ref` on it, so you have to copy each element.